### PR TITLE
Fixes for thread safety of CCDB manager and MW2 CDB tests

### PR DESCRIPTION
**Pull request by shahor02**

### CODE REVIEW

File: Detectors/Calibration/testMacros/retrieveFromCCDB.cxx [hunk 1]
```diff
@@ -23,7 +23,7 @@ bool initOptionsAndParse(bpo::options_description& options, int argc, char* argv
   options.add_options()(
     "TFs-in-parallel,m", bpo::value<int>()->default_value(8), "Number of TFs to simulate that access the CCDB in parallel")(
     "TF-processing-time,t", bpo::value<float>()->default_value(10.), "Seconds supposed to be needed to process a TF")(
-    "ccdb-sercer,s", bpo::value<std::string>()->default_value("ccdb-test.cern.ch:8080"), "CCDB server")(
+    "ccdb-sercer,s", bpo::value<std::string>()->default_value("http://ccdb-test.cern.ch:8080"), "CCDB server")(
```
Comment (chiarazampolli) [inline Detectors/Calibration/testMacros/retrieveFromCCDB.cxx:26]: Hi @shahor02 , can you fix the typo in the option? It should be "serVer".

File: CCDB/src/CcdbApi.cxx [hunk 2]
```diff
@@ -97,6 +100,7 @@ std::unique_ptr<std::vector<char>> CcdbApi::createObjectImage(const void* obj, s
 {
   // Create a binary image of the object, if CcdbObjectInfo pointer is provided, register there
   // the assigned object class name and the filename
+  std::lock_guard<std::mutex> guard(gIOMutex);
```
Comment (sawenzel) [inline CCDB/src/CcdbApi.cxx:103]: I am wonderning, why do you need to make CcdbApi threadsafe? This class is not a singleton and one could easily have a CccbApi instance per thread. At first, I understood that thread-safety is attacked via BasicCCDBManager but maybe this is not the case?

I am open to these changes but wonder if we need to pay the price of mutexes all the time.

Comment (Barthelemy) [reply CCDB/src/CcdbApi.cxx:103]: I agree with Sandro. Although speed is not the most relevant feature of the CcdbApi (it is slow due to protocol), I think that it was designed with the idea that you can have as many instances as you need. In QC we have 2 for example, one for the QCDB and one for the CCDB.

**Comment by shahor02** [comment]
@Barthelemy , @sawenzel : root global variables are shared among the threads, even if we create an instance of CcdbApi per thread, the storage and retrieval will not be thread safe (that's where I've started). 1st I had the mutex guards around the CcdbApi calls in workflow directly, but in this way we are penalized also by the network latency, so, I've minimized the overhead by grouping together and protecting unsafe root operations in CcdbApi.
For the BasicCCDBManager: with the cached enabled it is not thread safe due to the cache ownership, so I've introduced its intermediate non-singleton version to allow 1 manager per thread (or pool of managers to reuse them and profit from cache).

**Comment by sawenzel** [comment]
@shahor02 : Did you try enabling thread-safety in ROOT with `ROOT::EnableThreadSafety`?

**Comment by shahor02** [comment]
@sawenzel did not know about this EnableThreadSafety, indeed, it helps. Do we want to enable it globally ?

**Comment by davidrohr** [comment]
Where would you want to enable it globally?
For all of O2? Not sure, might have a performance impact, that's why it is optional I assume.

**Comment by shahor02** [comment]
I also think so, but then we have to protect unsafe operations locally, i.e. in the CcdbApi

**Comment by shahor02** [comment]
@sawenzel @Barthelemy any comments on this PR? It is needed for the MW2, so should go to the tag (presumably of today)

**Comment by sawenzel** [comment]
For me it's ok to continue like it is.

**Pull request merged**